scripts/yyVariable.js

```diff
@@ -526,37 +561,46 @@ function array_create( _size, _val )
 
 function array_copy( _dest, _dest_index, _src, _src_index, _length)
 {
-    if (_src_index < 0)
-    {
-        _length += _src_index;      /* Negative _src_index decrements _length... */
-        _dest_index -= _src_index;  /* ...increments _dest_index... */
-        _src_index = 0;             /* ...and finally gets clamped to zero. */
-    }
-
-    if (_length < 0) {
-        _length = 0; /* Negative _length gets clamped to zero. */
-    }
-
     if (Array.isArray(_dest) && Array.isArray(_src)) {
 
         _dest_index = yyGetInt32(_dest_index);
-        _src_index = yyGetInt32(_src_index);
-        _length = yyGetInt32( _length );
-
-        // fill any null elements prior to the copy with 0 (real) values.
-        // This mimics the initialisation flow of array data on VM/YYC when MemoryManager::SetLength is called.
-        if (_dest.length < _dest_index) {
-            for (var index = _dest.length; index < _dest_index + min(_src.length - _src_index, _length) ; ++index) {
-                if (_dest[index] == null) {
-                    _dest[index] = 0;
-                }
-            } // end for 
-        } // end if
+        _src_index = isFinite(_src_index) ? yyGetInt32(_src_index) : yyGetReal(_src_index);
+        _length = isFinite(_length) ? yyGetInt32(_length) : yyGetReal(_length);
+
+        // Compute the correct values for offset, loops and step (direction)
+        [ _src_index, _length, _step ] = computeIterationValues(_src.length, _src_index, _length);
 
-        for (var n = 0; n < _length && (_src_index + n) < _src.length; ++n) {
-            _dest[_dest_index + n] = _src[_src_index + n];
-        } // end for 
+        if (_length == 0) return;
+        
+        // Converts negative offset to positive value and clamps between allowed values
+        _dest_index = Math.max(_dest_index, -_dest.length);
+        if (_dest_index < 0) _dest_index = _dest.length + _dest_index;
+
+        // Fill any null elements prior to the copy with 0 (real) values. 
+        // This mimics the initialisation flow of array data on VM/YYC when MemoryManager::SetLength is called. 
+        if (_dest.length < _dest_index) { 
+            for (var index = _dest.length; index < _dest_index + min(_src.length - _src_index, _length) ; ++index) { 
+                if (_dest[index] == null) { 
+                    _dest[index] = 0; 
+                }
+            } // end for  
+        } // end if 
+
+        // Reverse indices we want to start at the end and go towards the beginning
+        _src_index = _src_index + ((_length - 1) * _step);
+        _dest_index += _length;
+        _step = -_step;
+
+        // Iterate over the input array range
+        while (_length > 0) {
+            _dest[--_dest_index] = _src[_src_index];
+            _src_index += _step;
+            _length -= 1;
+        }
     } // endif
+    else {
+        yyError( "argument0 and argument2 are not arrays");

```

Comment:
can you add Error: array_copy: or something so that we can know which function is throwing the error - when obfuscated you can't tell. And the message should be argument0 and/or argument2 is not an array?